triedb/firewood/account_trie.go

```diff
@@ -0,0 +1,275 @@
+//  2. The `Hash` method actually creates the proposal, since Firewood cannot calculate
+//     the hash of the trie without committing it. It is immediately dropped, and this
+//     can likely be optimized.
+type AccountTrie struct {
+	fw           *Database
+	parentRoot   common.Hash
+	root         common.Hash
+	reader       database.Reader
+	dirtyKeys    map[string][]byte // Store dirty changes
+	updateKeys   [][]byte
+	updateValues [][]byte
+	hasChanges   bool
+}
+
+func NewAccountTrie(root common.Hash, db *Database) (*AccountTrie, error) {
+	reader, err := db.Reader(root)
+	if err != nil {
+		return nil, err
+	}
+	return &AccountTrie{
+		fw:         db,
+		parentRoot: root,
+		reader:     reader,
+		dirtyKeys:  make(map[string][]byte),
+		hasChanges: true, // Start with hasChanges true to allow computing the proposal hash
+	}, nil
+}
+
+// GetAccount implements state.Trie.
+func (a *AccountTrie) GetAccount(addr common.Address) (*types.StateAccount, error) {
+	key := crypto.Keccak256Hash(addr.Bytes()).Bytes()
+
+	// First check if there's a pending update for this account
+	keyStr := string(key)
+	if updateValue, exists := a.dirtyKeys[keyStr]; exists {
+		// If the value is empty, it indicates deletion
+		if len(updateValue) == 0 {
+			return nil, nil
+		}
+		// Decode and return the updated account
+		account := new(types.StateAccount)
+		err := rlp.DecodeBytes(updateValue, account)
+		return account, err
+	}
+
+	// No pending update found, read from the underlying reader
+	accountBytes, err := a.reader.Node(common.Hash{}, key, common.Hash{})
+	if err != nil {
+		return nil, err
+	}
+
+	if accountBytes == nil {
+		return nil, nil
+	}
+
+	// Decode the account node
+	account := new(types.StateAccount)
+	err = rlp.DecodeBytes(accountBytes, account)
+	return account, err
+}
+
+// GetStorage implements state.Trie.
+func (a *AccountTrie) GetStorage(addr common.Address, key []byte) ([]byte, error) {
+	var combinedKey [2 * common.HashLength]byte
+	accountKey := crypto.Keccak256Hash(addr.Bytes()).Bytes()
+	storageKey := crypto.Keccak256Hash(key).Bytes()
+	copy(combinedKey[:common.HashLength], accountKey)
+	copy(combinedKey[common.HashLength:], storageKey)
+
+	// If the account has been deleted, we should return nil
+	if val, exists := a.dirtyKeys[string(accountKey)]; exists && len(val) == 0 {
+		return nil, nil
+	}
+
+	// Check if there's a pending update for this storage slot
+	keyStr := string(combinedKey[:])
+	if updateValue, exists := a.dirtyKeys[keyStr]; exists {
+		// If the value is empty, it indicates deletion
+		if len(updateValue) == 0 {
+			return nil, nil
+		}
+		// Decode and return the updated storage value
+		_, decoded, _, err := rlp.Split(updateValue)
+		return decoded, err
+	}
+
+	// No pending update found, read from the underlying reader
+	storageBytes, err := a.reader.Node(common.Hash{}, combinedKey[:], common.Hash{})
+	if err != nil {
+		return nil, err
+	}
+	if storageBytes == nil {
+		return nil, nil
+	}
```

Comment:
I guess this could be:
```suggestion
	if err != nil || storageBytes == nil {
		return nil, err
	}
```
at the cost of some readability